src/AI/JakeMan.java

```diff
@@ -785,16 +867,188 @@ private Map<XYCoord, UnitModel> queueUnitProductionActions(GameMap gameMap)
       log("No properties available to build.");
       return builds;
     }
+    int budget = myArmy.money;
+
+    // Fill out production with inf first, to trim the budget
+    ArrayList<MapLocation> infBases = CPI.getAllFacilitiesFor(infantry);
+    for( MapLocation loc : infBases )
+    {
+      Commander buyer = loc.getOwner();
+      final XYCoord coord = loc.getCoordinates();
+      final int cost = buyer.getBuyCost(infantry, coord);
 
+      if( cost <= budget )
+      {
+        builds.put(coord, infantry);
+        budget -= cost;
+      }
+      else
+        budget = -1;
+    }
+    if( budget < 3000 ) // arbitrary
+      return builds;
     log("Evaluating Production needs");
-    ArrayList<UnitModel> wantedTypes = new ArrayList<>();
 
+    ArrayList<FactoryThreatState> potentialMdBuilds = new ArrayList<>();
+    ArrayList<MapLocation> mdBases = CPI.getAllFacilitiesFor(mdTank);
+    UnitContext theVeh = new UnitContext(myArmy.cos[0], tank); // for getting the zone of concern
+    for( MapLocation loc : mdBases )
+      potentialMdBuilds.add(new FactoryThreatState(this, loc.getCoordinates(), gameMap, theVeh));
+
+    // "Net health" of the units I might want to build counters for, or use as counters.
+    HashMap<UnitModel, Integer> niceHealth = new HashMap<>();
+    HashMap<UnitModel, Integer> meanHealth = new HashMap<>();
```

Comment:
[nice/mean](https://github.com/timvieira/justified-variables)?

Reply:
Yessir.

Reply:
I guess I could go evil/good, but I feel like that's a little strong.